sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala

```diff
@@ -70,6 +74,10 @@ abstract class PartitioningAwareFileIndex(
     val selectedPartitions = if (partitionSpec().partitionColumns.isEmpty) {
       PartitionDirectory(InternalRow.empty, allFiles().filter(isNonEmptyFile)) :: Nil
     } else {
+      if (recursiveFileLookup) {
```

Comment:
This branch seems not reachable. Should we simply use `assert` here?

Reply:
Oh, it is reachable I think.
See `class PrunedInMemoryFileIndex` which explicitly set `partitionSpec`.